### ISSUE TITLE
Dependency resolution for 3rd party packages

### ISSUE THREAD

**Issue by dalcde** [iodide-project/pyodide#1100]
It would be nice for third party packages to be able to specify dependencies. Ideally, we can stop using `packages.json` for this (we might still need it for imports though).

One issue is that currently, we find all recursive dependencies, and then load them all at once. If we have to load a package before knowing its dependencies, this could end up being quite slower.

**Comment by rth** [comment]
> If we have to load a package before knowing its dependencies, this could end up being quite slower.

Yes, that is a major point. Also I find it much easier to debug and manipulate a single `packages.json` for dependencies, as opposed to having them scattered though different package files.

One way to address this use case, would be to support multiple channels. So we could set,
```js
windows.languagePluginUrl = [
    "https://cdn.jsdelivr.com/pyodide/...",
    "some-custom-url",
    ...
]
```
where each channel has its own `packages.json`, and at startup we load them and build a single dependency graph from it. Channels could have different priority, so that if `numpy` is defined in both the one from the top of the list is used.

**Comment by hoodmane** [comment]
I am not 100% sure what is proposed here but I think it has not been done yet. Perhaps @joemarshall or @ryanking13 will have some comments. I think having an API for updating the package lock at runtime might be helpful.

**Comment by ryanking13** [comment]
I think this issue is about adding an option to specify dependencies of third-party packages that are not in repodata.json.

For now, I think this is (indirectly) available by:

1. micropip.freeze(): Load packages once, then create a dependency tree from it.
2. micropip.install(..., deps=False): List all packages to be installed.
